File: src/LinkedList/BrowserHistory.java
package LinkedList;
public class BrowserHistory {
    ListNode head;
    ListNode current;
    ListNode previous;

    public BrowserHistory(String homepage){
        head = new ListNode(homepage);
        current = head;
    }

    public void visit(String url){
        current.next = new ListNode(url);
        previous = current;
        current = current.next;
        current.prev = previous;
    }

    public String back(int steps){
        for(int i = 0; i < steps; i++){
            if (current.prev != null){
                current = current.prev;
            }
        }
        return current.url;
    }

    public String forward(int steps){
        for(int i = 0; i < steps; i++){
            if (current.next != null){
                current = current.next;
            }
        }
        return current.url;
    }

    private class ListNode {
        String url;
        ListNode next;
        ListNode prev;
    
        ListNode(String url){
            this.url = url;
            this.next = null;
            this.prev = null;
        }
    }
}

File: src/Python/LinkedList/browser_history.py
class ListNode:
    def __init__(self, page: str, next = None, prev = None):
        self.page = page
        self.next = next
        self.prev = prev

class browserHistory:
    def __init__(self, homepage: str):
        home = ListNode(homepage)
        self.home = home
        self.curr = home

    def visit(self, url: str) -> None:
        tmp = ListNode(url)
        self.curr.next  = tmp
        tmp.prev = self.curr
        self.curr = tmp

    def back(self, steps: int) -> str:
        while self.curr.prev and steps > 0:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.page

    def forward(self, steps: int) -> str:
        while self.curr.next and steps > 0:
            self.curr = self.curr.next
            steps -= 1
        return self.curr.page